botocore/docs/bcdoc/docstringparser.py

```diff
@@ -22,30 +21,175 @@ class DocStringParser(html_parser.HTMLParser):
     """
 
     def __init__(self, doc):
-        html_parser.HTMLParser.__init__(self)
+        six.moves.html_parser.HTMLParser.__init__(self)
         self.doc = doc
-        self.unhandled_tags = []
+        self.tree = HTMLTree(self.doc)
 
     def handle_starttag(self, tag, attrs):
-        handler_name = 'start_%s' % tag
-        if hasattr(self.doc.style, handler_name):
-            getattr(self.doc.style, handler_name)(attrs)
-        else:
-            self.unhandled_tags.append(tag)
+        self.tree.add_tag(tag, attrs=attrs)
 
     def handle_endtag(self, tag):
-        handler_name = 'end_%s' % tag
-        if hasattr(self.doc.style, handler_name):
-            getattr(self.doc.style, handler_name)()
+        self.tree.add_tag(tag, is_start=False)
 
     def handle_data(self, data):
-        if data.isspace():
-            data = ' '
+        self.tree.add_data(data)
+
+    def close(self):
+        # HTMLParser is an old style class, so the super() method will not work.
+        six.moves.html_parser.HTMLParser.close(self)
+        self.tree.write()
```

Comment:
It's a little unfortunate that we have to buffer the entire tree and write it out on the ``close()`` as opposed to before we actually just wrote out with the style as we were feeding the html. I understand the need for an abstraction on top of the style but would it be possible to write out the data to the doc as you add the children? I think it may be possible, but I do not think it is a dealbreaker if we are unable to it because we only invoke the html parser feed on docstings for single shapes and members.